### PR TITLE
param `auto_init` for repo and fix race conditions for branch protection

### PR DESCRIPTION
**Pull request by majormoses**
## Description

Basically when you create a new repo you can have a race condition where it wants to create the branch protection. The problem is that you have two dependencies:
- the repo itself
- the branch to be protected needs to exist

By adding an explicit depends we can ensure that only after the repo resource has been run will it attempt to setup branch protections.

Also in the case of wanting to create a repository and copy the history of another project in you can specify `auto_init` and it will not attempt to initialize the repo with a commit or setup protection. After copying the history you can flip `auto_init` to `true` which will flip the branch protection on without creating any commits.

Signed-off-by: Ben Abrams <me@benabrams.it>

Additional scope after resurrecting this pull request:
- check if a repo has been archived and if so set `pull` permission to prevent terraform from thinking that permissions need to be updated. Otherwise once a repository has been archived it would either require updating those vars to pull but it seems easier to just check the condition and act in the users best interest.

### Issues Resolved

[List any existing issues this PR resolves]

### Check List
- [x] All tests pass. See https://github.com/sous-chefs/apache2/blob/master/TESTING.md
~- [ ] New functionality includes testing.~
~- [ ] New functionality has been documented in the README if applicable~

### CODE REVIEW

**Comment by majormoses** [comment]
Hmm I need to look into why it thinks that should force a destruction (adding `auto_init`) might be version of module we are pinned at.

**Comment by majormoses** [comment]
might need to open up an issue on upstream as that should not be happening.

**Comment by damacus** [comment]
Yeah I saw this with this particular provider.

File: modules/repository/main.tf [hunk 1]
```diff
@@ -13,13 +13,30 @@ resource "github_repository" "repository" {
...
+  # creation from github api side but we also use it to determine if branch
+  # protections should be enabled.
+  auto_init = "${var.auto_init}"
 }
 
 resource "github_branch_protection" "repository_master" {
+  # this is a bit of a hack to allow people to create repositories uninitialized
+  # and then add branch protection later. The use cases are mostly around needing
+  # to create "forked" private repositories
+  count = "${var.auto_init}"
```
Comment (damacus) [inline modules/repository/main.tf]: Nice hack

File: modules/repository/variables.tf [hunk 2]
```diff
@@ -78,3 +78,20 @@ variable "additional_topics" {
   type    = "list"
   default = []
 }
+
+variable "auto_init" {
+  default = true
```
Comment (damacus) [inline modules/repository/variables.tf]: You're probably better off just sticking with the slightly less verbose 
https://www.terraform.io/docs/providers/github/r/repository.html#auto_init

**Comment by damacus** [comment]
Closing this for now

**Comment by majormoses** [comment]
@damacus the only changes I see in my plan are completely unrelated. It looks like https://github.com/sous-chefs/terraform-github-org/commit/41fc86e18cf2fe794efc26d537e16adf46a274bd archived the repo but the module wants to reset the permissions.

**Comment by majormoses** [comment]
Got a quick fix in for the permissions "conflict" when archiving a repo and the only changes left are adding Julian to the org and team. Was he potentially accidentally removed from the UI?
- https://github.com/sous-chefs/terraform-github-org/blob/85928326fa7278c6b8c40f877815626a2173db71/modules/org_membership/main.tf#L66-L69
- https://github.com/sous-chefs/terraform-github-org/blob/85928326fa7278c6b8c40f877815626a2173db71/chef_de_partie.tf#L74-L78

File: modules/repository/main.tf [hunk 3]
```diff
@@ -36,11 +61,11 @@ resource "github_branch_protection" "repository_master" {
 resource "github_team_repository" "repository_everyone" {
   team_id    = "${var.chef_de_partie}"
   repository = "${github_repository.repository.name}"
-  permission = "${var.everyone_permission}"
+  permission = "${var.archived ? "pull" : var.everyone_permission}"
```
Comment (majormoses) [inline modules/repository/main.tf]: I thought about this and we could instead set the count to `0` if thats set but that removes the access altogether which might be OK for a public repository and would remove explicit permissions in favor of implicit pull permissions from a public repository. I am not sure if this is a concern for private repositories but I went with the slightly more verbose option for now. I can also comment this if you think if the intent is not obvious.

**Comment by damacus** [comment]
I think Julian removed himself when he left Chef, and this is probably the first plan to pick that change up.

**Pull request merged**